File: Tema 4. Modelo de programación lógica/sistemaRiego.pl
% Módulo de sensores (datos dinámicos)
:- dynamic humedad_suelo/1.
:- dynamic temperatura/1.
:- dynamic hora/1.
:- dynamic pronostico_lluvia/1.

% Valores iniciales de los sensores
humedad_suelo(baja).
temperatura(35).
hora(20).
pronostico_lluvia(false).

% Tipos de humedad posibles
humedad_posible(baja).
humedad_posible(media).
humedad_posible(alta).

% Rangos válidos para sensores
rango_temperatura_valida(T) :- number(T), T >= -10, T =< 50.
rango_hora_valida(H) :- integer(H), H >= 0, H =< 23.

% Validación de datos de sensores
sensor_valido :-
    humedad_suelo(H), humedad_posible(H),
    temperatura(T), rango_temperatura_valida(T),
    hora(Hora), rango_hora_valida(Hora),
    pronostico_lluvia(P), (P == true ; P == false).

% Reglas mejoradas para hora adecuada
hora_adecuada :- hora(H), (H < 10 ; H > 18).  % Ahora solo riego nocturno para mejor eficiencia
    hora_adecuada :- hora(H), (H < 10 ; H > 18).

% Regla principal mejorada con validación
activar_riego :- 
    sensor_valido,
    humedad_suelo(baja), 
    pronostico_lluvia(false), 
    hora_adecuada,
    \+ condiciones_peligrosas.

% Condiciones que impiden el riego
condiciones_peligrosas :-
    temperatura(T), T >= 40.  % Temperatura extremadamente alta

% Estados del sistema mejorados
estado_riego('Activado - Riego en curso') :- activar_riego.
estado_riego('No activado - Humedad adecuada') :- 
    humedad_suelo(media), writeln('Info: Humedad del suelo en nivel óptimo').
estado_riego('No activado - Humedad alta') :- 
    humedad_suelo(alta), writeln('Advertencia: Suelo ya saturado de agua').
estado_riego('No activado - Hora inadecuada') :- 
    \+ hora_adecuada, writeln('Info: Esperando horario nocturno para riego').
estado_riego('No activado - Lluvia pronosticada') :-
    pronostico_lluvia(true), writeln('Info: Se espera lluvia pronto').
estado_riego('No activado - Condiciones peligrosas') :-
    condiciones_peligrosas, writeln('ALERTA: Condiciones extremas detectadas').

% Sistema de alertas mejorado
alerta_calor :- 
    temperatura(T), 
    T >= 32,
    writeln('Alerta: Temperatura elevada detectada').

alerta_helada :-
    temperatura(T),
    T =< 2,
    writeln('Alerta: Riesgo de heladas detectado').

% Recomendaciones avanzadas
recomendacion :-
    activar_riego,
    alerta_calor, !,
    writeln('Recomendación: Riego activado con temperatura alta.'),
    writeln('* Usar riego por goteo para eficiencia hídrica'),
    writeln('* Verificar aspersores para evitar evaporación').

recomendacion :-
    activar_riego,
    writeln('Recomendación: Riego activado bajo condiciones óptimas').

recomendacion :-
    condiciones_peligrosas,
    writeln('Recomendación: Suspender riego por condiciones extremas').

recomendacion :-
    humedad_suelo(alta),
    writeln('Recomendación: Reducir frecuencia de riego - suelo saturado').

recomendacion :-
    writeln('Recomendación: Monitoreo continuo - sin acciones requeridas').

% Interface de usuario
mostrar_estado :-
    writeln('\n=== ESTADO DEL SISTEMA DE RIEGO ==='),
    estado_riego(Estado),
    writeln('Estado: '), writeln(Estado),
    (alerta_calor ; alerta_helada ; true),
    recomendacion,
    writeln('==================================\n').

% Predicados para actualizar sensores
actualizar_humedad(Valor) :-
    retract(humedad_suelo(_)),
    asserta(humedad_suelo(Valor)),
    writeln('Humedad actualizada').

actualizar_temperatura(Valor) :-
    number(Valor),
    retract(temperatura(_)),
    asserta(temperatura(Valor)),
    writeln('Temperatura actualizada').

actualizar_hora(Valor) :-
    integer(Valor),
    Valor >= 0, Valor =< 23,
    retract(hora(_)),
    asserta(hora(Valor)),
    writeln('Hora actualizada').

actualizar_pronostico(Valor) :-
    (Valor == true ; Valor == false),
    retract(pronostico_lluvia(_)),
    asserta(pronostico_lluvia(Valor)),
    writeln('Pronóstico actualizado').

File: Tema 4. Modelo de programación lógica/sistemaRiego.py
class SistemaRiego:
    def __init__(self):
        # Datos dinámicos de los sensores
        self.humedad_suelo = "baja"
        self.temperatura = 35
        self.hora = 20
        self.pronostico_lluvia = False
        
        # Tipos de humedad posibles
        self.humedades_posibles = ["baja", "media", "alta"]
    
    def rango_temperatura_valida(self, t):
        return isinstance(t, (int, float)) and -10 <= t <= 50
    
    def rango_hora_valida(self, h):
        return isinstance(h, int) and 0 <= h <= 23
    
    def sensor_valido(self):
        return (self.humedad_suelo in self.humedades_posibles and
                self.rango_temperatura_valida(self.temperatura) and
                self.rango_hora_valida(self.hora) and
                isinstance(self.pronostico_lluvia, bool))
    
    def hora_adecuada(self):
        return self.hora < 10 or self.hora > 18
    
    def condiciones_peligrosas(self):
        return self.temperatura >= 40
    
    def activar_riego(self):
        return (self.sensor_valido() and
                self.humedad_suelo == "baja" and
                not self.pronostico_lluvia and
                self.hora_adecuada() and
                not self.condiciones_peligrosas())
    
    def estado_riego(self):
        if self.activar_riego():
            return "Activado - Riego en curso"
        elif self.humedad_suelo == "media":
            print("Info: Humedad del suelo en nivel óptimo")
            return "No activado - Humedad adecuada"
        elif self.humedad_suelo == "alta":
            print("Advertencia: Suelo ya saturado de agua")
            return "No activado - Humedad alta"
        elif not self.hora_adecuada():
            print("Info: Esperando horario nocturno para riego")
            return "No activado - Hora inadecuada"
        elif self.pronostico_lluvia:
            print("Info: Se espera lluvia pronto")
            return "No activado - Lluvia pronosticada"
        elif self.condiciones_peligrosas():
            print("ALERTA: Condiciones extremas detectadas")
            return "No activado - Condiciones peligrosas"
        else:
            return "Estado desconocido"
    
    def alerta_calor(self):
        if self.temperatura >= 32:
            print("Alerta: Temperatura elevada detectada")
            return True
        return False
    
    def alerta_helada(self):
        if self.temperatura <= 2:
            print("Alerta: Riesgo de heladas detectado")
            return True
        return False
    
    def recomendacion(self):
        if self.activar_riego() and self.alerta_calor():
            print("Recomendación: Riego activado con temperatura alta.")
            print("* Usar riego por goteo para eficiencia hídrica")
            print("* Verificar aspersores para evitar evaporación")
        elif self.activar_riego():
            print("Recomendación: Riego activado bajo condiciones óptimas")
        elif self.condiciones_peligrosas():
            print("Recomendación: Suspender riego por condiciones extremas")
        elif self.humedad_suelo == "alta":
            print("Recomendación: Reducir frecuencia de riego - suelo saturado")
        else:
            print("Recomendación: Monitoreo continuo - sin acciones requeridas")
    
    def mostrar_estado(self):
        print("\n=== ESTADO DEL SISTEMA DE RIEGO ===")
        print("Estado:", self.estado_riego())
        self.alerta_calor() or self.alerta_helada()
        self.recomendacion()
        print("==================================\n")
    
    # Métodos para actualizar sensores
    def actualizar_humedad(self, valor):
        if valor in self.humedades_posibles:
            self.humedad_suelo = valor
            print("Humedad actualizada")
        else:
            print(f"Error: Humedad {valor} no válida. Valores posibles: {self.humedades_posibles}")
    
    def actualizar_temperatura(self, valor):
        if self.rango_temperatura_valida(valor):
            self.temperatura = valor
            print("Temperatura actualizada")
        else:
            print(f"Error: Temperatura {valor} fuera de rango válido (-10 a 50)")
    
    def actualizar_hora(self, valor):
        if self.rango_hora_valida(valor):
            self.hora = valor
            print("Hora actualizada")
        else:
            print(f"Error: Hora {valor} no válida (0-23)")
    
    def actualizar_pronostico(self, valor):
        if isinstance(valor, bool):
            self.pronostico_lluvia = valor
            print("Pronóstico actualizado")
        else:
            print("Error: El pronóstico debe ser True o False")

# Ejemplo de uso
if __name__ == "__main__":
    sistema = SistemaRiego()
    
    # Mostrar estado inicial
    sistema.mostrar_estado()
    
    # Actualizar algunos valores
    sistema.actualizar_humedad("media")
    sistema.actualizar_temperatura(38)
    sistema.actualizar_hora(21)
    
    # Mostrar estado después de cambios
    sistema.mostrar_estado()
    
    # Probar condiciones de riego
    sistema.actualizar_humedad("baja")
    sistema.actualizar_temperatura(25)
    sistema.actualizar_pronostico(False)
    sistema.mostrar_estado()